01_explore_data/nbconverted/count_all_genes_filters.py

```diff
@@ -0,0 +1,193 @@
+#!/usr/bin/env python
+# coding: utf-8
+
+# ## Count number of valid genes after applying cancer type filters
+# 
+# In [the preprocessing code for our classifiers](https://github.com/greenelab/mpmp/blob/5d5fa0823b00fc3080d3a9db69d8d6704f554549/mpmp/utilities/tcga_utilities.py#L84), we filter out cancer types that don't contain at least 5% of samples mutated and at least 10 total samples mutated, for a given target gene.
+# 
+# We were curious how many total genes these filters would give us, if we look at _all_ ~20,000 genes we have mutation data for. This script filters samples for each gene and counts the number of samples/cancer types that would be included in our classifiers.
+
+# In[1]:
+
+
+import sys
+from pathlib import Path
+
+import numpy as np
+import pandas as pd
+
+import mpmp.config as cfg
+from mpmp.data_models.tcga_data_model import TCGADataModel
+import mpmp.utilities.data_utilities as du
+
+
+# ### Choose whether or not to use CNV info
+# 
+# Normally we'd use copy gain for oncogenes and copy loss for TSGs, but since we don't have these annotations for all the genes in the genome, there are two ways to handle marking samples as "mutated":
+# 
+# 1. Just use point mutations for all genes, no copy number info
+# 2. Use point mutations and copy gain/loss for all genes
+# 
+# We provide an option here to run either of these.
+
+# In[2]:
+
+
+use_copy_data = False
+
+if use_copy_data:
+    output_file = Path('./gene_sample_count_with_copy.tsv')
+else:
+    output_file = Path('./gene_sample_count_no_copy.tsv')
+
+
+# In[3]:
+
+
+# load sample info and mutation data, this takes some time
+tcga_data = TCGADataModel(seed=cfg.default_seed, verbose=False)
+pancancer_data = du.load_pancancer_data()
+
+
+# In[4]:
+
+
+def gene_sample_count(gene, data_model, classification='neither'):
+    """Count valid samples/cancer types for a given gene."""
+    try:
+        tcga_data.process_data_for_gene(gene,
+                                        classification,
+                                        None)
+        sample_count = tcga_data.X_df.shape[0]
+        cancer_types = ';'.join(tcga_data.sample_info_df
+          .loc[tcga_data.X_df.index, 'cancer_type']
+          .unique()
+        )
+    except KeyError:
+        sample_count = np.nan
+        cancer_types = pd.NA
+        
+    return (gene, sample_count, cancer_types)
+
+
+# In[5]:
+
+
+# cache partial results and load them
+# we're running this script for 20,000 genes, so it's nice to save progress in case
+# execution is interrupted
```

Comment:
Is there a scenario where execution fails partway through a gene and so some information for that single gene gets written but not all?

Reply:
I don't think that should happen - in [the code here](https://github.com/greenelab/mpmp/blob/46b5549edf8a1a6e5499d86050fff5dc70f2756d/01_explore_data/nbconverted/count_all_genes_filters.py#L110), I'm only saving the information every `save_every` iterations, so if execution fails partway through a gene we'd just go back to the previous iteration. We might have to rerun a few genes that way, but that's fine assuming `save_every` is fairly small (on the order of tens or hundreds).